block/stateupdate_validator.go

```diff
@@ -0,0 +1,162 @@
+
+	// validate P2P blocks against DA blocks
+	err = v.ValidateP2PBlocks(daBlocks, p2pBlocks)
+	if err != nil {
+		return err
+	}
+
+	// update the last validated height to the batch last block height
+	v.blockManager.UpdateLastValidatedHeight(batch.EndHeight)
+
+	return nil
+}
+
+// ValidateP2PBlocks basically compares that the blocks applied from P2P are the same blocks included in the batch and retrieved from DA.
+// Since DA blocks have been already validated against Hub state info block descriptors, if P2P blocks match with DA blocks, it means they are also validated against state info block descriptors.
+func (v *StateUpdateValidator) ValidateP2PBlocks(daBlocks []*types.Block, p2pBlocks map[uint64]*types.Block) error {
+	// nothing to compare
+	if len(p2pBlocks) == 0 {
+		return nil
+	}
+
+	// iterate over daBlocks and compare hashes with the corresponding block from P2P (if exists) to see whether they are actually the same block
+	for _, daBlock := range daBlocks {
+
+		p2pBlock, ok := p2pBlocks[daBlock.Header.Height]
+		if !ok {
+			continue
+		}
+		p2pBlockHash, err := blockHash(p2pBlock)
+		if err != nil {
+			return err
+		}
+		daBlockHash, err := blockHash(daBlock)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(p2pBlockHash, daBlockHash) {
+			return types.NewErrStateUpdateDoubleSigningFraud(daBlock.Header.Height, p2pBlock)
+		}
+
+	}
+	return nil
+}
+
+// ValidateDaBlocks checks that the information included in the Hub state info (height, state roots and timestamps), correspond to the blocks obtained from DA.
+func (v *StateUpdateValidator) ValidateDaBlocks(slBatch *settlement.ResultRetrieveBatch, daBlocks []*types.Block) error {
+	// we first verify the numblocks included in the state info match the block descriptors and the blocks obtained from DA
+	numSlBDs := uint64(len(slBatch.BlockDescriptors))
+	numDABlocks := uint64(len(daBlocks))
+	numSLBlocks := slBatch.NumBlocks
+	if numSLBlocks != numDABlocks || numSLBlocks != numSlBDs {
+		return types.NewErrStateUpdateNumBlocksNotMatchingFraud(slBatch.EndHeight, numSLBlocks, numDABlocks, numSLBlocks)
+	}
+
+	// we compare all DA blocks against the information included in the state info block descriptors
+	for i, bd := range slBatch.BlockDescriptors {
+		// height check
+		if bd.Height != daBlocks[i].Header.Height {
+			return types.NewErrStateUpdateHeightNotMatchingFraud(slBatch.StateIndex, bd.Height, daBlocks[i].Header.Height)
+		}
+		// we compare the state root between SL state info and DA block
+		if !bytes.Equal(bd.StateRoot, daBlocks[i].Header.AppHash[:]) {
+			return types.NewErrStateUpdateStateRootNotMatchingFraud(slBatch.StateIndex, bd.Height, bd.StateRoot, daBlocks[i].Header.AppHash[:])
+		}
+
+		// we compare the timestamp between SL state info and DA block
+		if !bd.Timestamp.Equal(daBlocks[i].Header.GetTimestamp()) {
+			return types.NewErrStateUpdateTimestampNotMatchingFraud(slBatch.StateIndex, bd.Height, bd.Timestamp, daBlocks[i].Header.GetTimestamp())
+		}
+	}
```

Comment:
and drs I guess

Reply:
drs is part of another pr. i removed any reference to it to avoid confusion